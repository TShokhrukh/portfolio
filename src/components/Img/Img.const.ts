import { cn } from '@bem-react/classname'

export const cnImg = cn('Img')
