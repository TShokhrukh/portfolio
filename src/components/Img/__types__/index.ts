export * from './Img.types'
export * from './Img_fill.types'
