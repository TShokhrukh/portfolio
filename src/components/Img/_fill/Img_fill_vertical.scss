.Img_fill_vertical {
  height: 100%;
}