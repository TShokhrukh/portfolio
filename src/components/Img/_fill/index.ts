export * from './Img_fill_horizontal'
export * from './Img_fill_vertical'
