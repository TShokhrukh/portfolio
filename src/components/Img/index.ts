export * from './build'
export * from './__types__'
export * from './_fill'
export { Img as ImgBase } from './Img'
