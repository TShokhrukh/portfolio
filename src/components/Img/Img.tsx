import React, { useState, FC } from 'react'
import { useIntersectionObserver } from '~/hooks/useIntersectionObserver'
import { cnImg } from './Img.const'

import plug from './res/plug.png'
import type { IImgProps } from './__types__'
import './Img.scss'

export const Img: FC<IImgProps> = ({
  className,
  src,
  plugSrc: preSrc = plug,
  alt = '...',
  showPlug = true,
  ...props
}) => {
  const [state, setState] = useState(() => !showPlug)
  const ref = useIntersectionObserver<HTMLImageElement>(() => {
    setState(true)
  })

  return (
    <img {...props} ref={ref}
      src={state ? src : preSrc} alt={alt}
      className={cnImg({}, [className])} />
  )
}
