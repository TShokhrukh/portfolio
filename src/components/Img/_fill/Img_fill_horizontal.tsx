import React from 'react'
import { withBemMod } from '@bem-react/core'
import { cnImg } from '../Img.const'
import type { IImgProps, IImgFillHorizontalProps } from '../__types__'
import './Img_fill_horizontal.scss'

export const withImgFillHorizontal = withBemMod<IImgFillHorizontalProps, IImgProps>(
  cnImg(),
  { fill: 'horizontal' },
  (Img) => function ImgFillHorizontal (props) {
    return <Img {...props} fill={undefined} />
  }
)
