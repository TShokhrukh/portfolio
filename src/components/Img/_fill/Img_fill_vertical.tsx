import React from 'react'
import { withBemMod } from '@bem-react/core'
import { cnImg } from '../Img.const'
import type { IImgProps, IImgFillVerticalProps } from '../__types__'
import './Img_fill_vertical.scss'

export const withImgFillVertical = withBemMod<IImgFillVerticalProps, IImgProps>(
  cnImg(),
  { fill: 'vertical' },
  (Img) => function ImgFillVertical (props) {
    return <Img {...props} fill={undefined} />
  }
)
