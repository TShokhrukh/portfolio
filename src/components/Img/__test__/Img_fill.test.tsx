/* global test, expect, describe */
import React from 'react'
import { compose, composeU } from '@bem-react/core'
import { render, screen } from '@testing-library/react'
import { Img as ImgBase } from '../Img'
import { withImgFillHorizontal, withImgFillVertical } from '../_fill'

const Img = compose(
  composeU(withImgFillHorizontal, withImgFillVertical)
)(ImgBase)

const fills = ['horizontal', 'vertical'] as const

describe.each(fills)('Img_fill_%s', (fill) => {
  test('must return the component template', () => {
    render(<Img data-testid='image' fill={fill} />)

    expect(screen.getByTestId('image')).toMatchSnapshot()
  })
})
