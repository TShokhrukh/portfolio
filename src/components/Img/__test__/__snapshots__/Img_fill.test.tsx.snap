// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Img_fill_horizontal must return the component template 1`] = `
<img
  alt="..."
  class="Img Img_fill_horizontal"
  data-testid="image"
/>
`;

exports[`Img_fill_vertical must return the component template 1`] = `
<img
  alt="..."
  class="Img Img_fill_vertical"
  data-testid="image"
/>
`;
